### PR TITLE
Fix clang warnings

### PR DESCRIPTION
**Pull request by samueljackson92**
**Description of work.**
Fixes clang warnings cropping up in clang `6.0.0` and unifies the warning suppression macros to work for both clang and gcc.

**To test:**

 - Mantid needs to compile without warnings on GCC and Clang builds.

_No associated issue_

Does this update require release notes?
- [ ] Yes
- [x] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by martyngigg — COMMENTED** [review]
A nice clean up of the warning macros.

Just a few in line comments.

File: Framework/API/test/ExperimentInfoTest.h [hunk 1]
```diff
@@ -156,14 +156,12 @@ class ExperimentInfoTest : public CxxTest::TestSuite {
 
   void test_GetSetSample() {
     ExperimentInfo ws;
-    TS_ASSERT(&ws.sample());
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/API/test/ExperimentInfoTest.h]: It seemed like an odd test in context. This produced a warning about addresses being convertible to bool, so I removed. I can re-add it an do a cast instead.

[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/API/test/ExperimentInfoTest.h]: Good point. The other tests in this method should essentially cover whether this object is valid. It's fine to leave it out.

File: Framework/Kernel/inc/MantidKernel/WarningSuppressions.h [hunk 2]
```diff
@@ -31,68 +31,88 @@
...
+#define CLANG_VERSION ((__clang_major__ * 100) + __clang_minor__)
+#endif
 
-// convenience for getting gcc version
+#if defined(__GNUC__) && !(defined(__INTEL_COMPILER)) && !defined(__clang__)
 #define GCC_VERSION                                                            \
   (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
+#endif
 
+// Currently this is only defined for gcc
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Kernel/inc/MantidKernel/WarningSuppressions.h]: No

File: Framework/Kernel/inc/MantidKernel/WarningSuppressions.h [hunk 3]
```diff
@@ -31,68 +31,88 @@
...
-// undefine definition from Poco 1.6
-#ifdef GCC_DIAG_JOINSTR
-#undef GCC_DIAG_JOINSTR
-#endif
-#define GCC_DIAG_JOINSTR(x, y) GCC_DIAG_STR(x##y)
-// undefine definition from Poco 1.6
-#ifdef GCC_DIAG_DO_PRAGMA
-#undef GCC_DIAG_DO_PRAGMA
+#define GNU_DIAG_MAKE_WARNING(x) "-W" x
+#define GNU_DIAG_DO_PRAGMA(x) _Pragma(#x)
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Kernel/inc/MantidKernel/WarningSuppressions.h]: It is not needed. I can remove it

File: Framework/PythonInterface/mantid/api/src/PythonAlgorithm/AlgorithmAdapter.cpp [hunk 4]
```diff
@@ -145,19 +146,24 @@ bool AlgorithmAdapter<BaseAlgorithm>::isRunning() const {
     return SuperClass::isRunning();
   } else {
     Environment::GlobalInterpreterLock gil;
+
+    GNU_DIAG_OFF("parentheses-equality")
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/PythonInterface/mantid/api/src/PythonAlgorithm/AlgorithmAdapter.cpp]: Yes. I believe so.

File: Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp [hunk 5]
```diff
@@ -58,25 +58,22 @@ using declarePropertyType3 = void (*)(boost::python::object &,
 using declarePropertyType4 = void (*)(boost::python::object &,
                                       const std::string &,
                                       const boost::python::object &, const int);
-#ifdef __clang__
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wunknown-pragmas"
-#pragma clang diagnostic ignored "-Wunused-local-typedef"
-#endif
+GNU_DIAG_OFF("unknown-pragmas")
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp]: Yes I think so.

File: Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp [hunk 6]
```diff
@@ -186,13 +187,16 @@ const std::string PropertyWithValueFactory::isArray(PyObject *const object) {
 
     PyObject *item = PySequence_Fast_GET_ITEM(object, 0);
     // Boolean can be cast to int, so check first.
+    GNU_DIAG_OFF("parentheses-equality")
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp]: Yes I think so

**Review by martyngigg — APPROVED** [review]
A great clean up of the warning macros. It nice that we now don't have to wrap certain macros with `clang-format off/on` statements!

**Pull request merged**